File: python/backends/ONNXRuntime/yolo_onnxruntime_pose.py
'''
Author: taifyang
Date: 2026-01-03 23:28:03
LastEditTime: 2026-01-20 22:09:31
Description: onnxruntime inference class for YOLO pose algorithm
'''


from backends.utils import *
from backends.yolo_pose import *
from backends.ONNXRuntime.yolo_onnxruntime import *


'''
description: onnxruntime inference class for the YOLO pose algorithm
'''
class YOLO_ONNXRuntime_Pose(YOLO_ONNXRuntime, YOLO_Pose):
    '''
    description:            construction method
    param {*} self          instance of class
    param {str} algo_type   algorithm type
    param {str} device_type device type
    param {str} model_type  model type
    param {str} model_path  model path
    return {*}
    '''   
    def __init__(self, algo_type:str, device_type:str, model_type:str, model_path:str) -> None:
        YOLO_ONNXRuntime.__init__(self, algo_type, device_type, model_type, model_path)
        YOLO_Pose.__init__(self, algo_type, device_type, model_type, model_path)

    '''
    description:    model pre-process
    param {*} self  instance of class
    return {*}
    '''    
    def pre_process(self) -> None:
        input = letterbox(self.image, self.inputs_shape)
        input = input[:, :, ::-1].transpose(2, 0, 1)  #BGR2RGB and HWC2CHW
        input = input / 255.0
        if self.model_type == 'FP32' or self.model_type == 'INT8':
            input = np.expand_dims(input, axis=0).astype(dtype=np.float32)
        elif self.model_type == 'FP16':
            input = np.expand_dims(input, axis=0).astype(dtype=np.float16)
        self.inputs[self.inputs_name[0]] = input
    
    '''
    description:    model post-process
    param {*} self  instance of class
    return {*}
    '''         
    def post_process(self) -> None:
        output = np.squeeze(self.outputs[0]).astype(dtype=np.float32) 
        if self.algo_type in ['YOLOv8', 'YOLOv11', 'YOLOv12']:
            scores = output[..., 4]
            xc = scores > self.score_threshold 
            output[..., :4] = xywh2xyxy(output[..., :4])
            box = output[xc][:, :4]
            scores =  np.expand_dims(scores[xc], axis=1)
            cls = np.zeros((len(box), 1))
            kpts = output[xc][..., 5:]
        elif self.algo_type in ['YOLO26']:
            output = output[output[..., 4] > self.score_threshold]
            box = output[:, :4]
            scores = output[..., 4:5]
            cls = output[..., 5:6]
            kpts = output[..., 6:]
        boxes = np.concatenate([box, scores, cls, kpts], axis=1)
        
        if len(boxes):  
            if self.algo_type in ['YOLOv8', 'YOLOv11', 'YOLOv12']: 
                indices = nms(boxes, scores, self.iou_threshold) 
                boxes = boxes[indices]
            boxes = scale_boxes(boxes, self.inputs_shape, self.image.shape)
            if self.render_result:
                self.result = self.draw_result(boxes)    

File: cpp/onnxruntime/yolo_onnxruntime_pose.cpp
/* 
 * @Author: taifyang
 * @Date: 2026-01-03 19:35:16
 * @LastEditTime: 2026-01-17 20:26:17
 * @Description: source file for YOLO onnxruntime pose
 */

 #include "yolo_onnxruntime.h"

void YOLO_ONNXRuntime_Pose::init(const Algo_Type algo_type, const Device_Type device_type, const Model_Type model_type, const std::string model_path)
{
	if (algo_type != YOLOv8 && algo_type != YOLOv11 && algo_type != YOLOv12 && algo_type != YOLO26)
	{
		std::cerr << "unsupported algo type!" << std::endl;
		std::exit(-1);
	}
	YOLO_ONNXRuntime::init(algo_type, device_type, model_type, model_path);
	YOLO_Pose::init(algo_type, device_type, model_type, model_path);

	m_output0.resize(m_output_numdet);
	if (m_model_type == FP16)
	{
		m_input_fp16.resize(m_input_numel);
		m_output0_fp16.resize(m_output_numdet);
	}
}

void YOLO_ONNXRuntime_Pose::pre_process()
{
	YOLO_ONNXRuntime_Detect::pre_process();
}

void YOLO_ONNXRuntime_Pose::process()
{
	YOLO_ONNXRuntime_Detect::process();
}

void YOLO_ONNXRuntime_Pose::post_process()
{
	std::vector<cv::Rect> boxes;
	std::vector<float> scores;
	std::vector<int> class_ids;
	std::vector<std::vector<float>> keypoints;

	for (int i = 0; i < m_output_numbox; ++i)
	{
		float* ptr = m_output0.data() + i * m_output_numprob;

		float score = ptr[4];
		if (score < m_score_threshold)
			continue;
		
		cv::Rect box;
		std::vector<float> keypoint(51);
		if (m_algo_type == YOLOv8 || m_algo_type == YOLOv11 || m_algo_type == YOLOv12)
		{
			float x = ptr[0];
			float y = ptr[1];
			float w = ptr[2];
			float h = ptr[3];
			int left = int(x - 0.5 * w) > 0 ? int(x - 0.5 * w) : 0;
			int top = int(y - 0.5 * h) > 0 ? int(y - 0.5 * h) : 0;
			int width = int(w) > 0 ? int(w) : 0;
			int height = int(h)> 0 ? int(h) : 0;
			width = (left + width) < m_image.cols ? width : (m_image.cols - left);
			height = (top + height) < m_image.rows ? height : (m_image.rows - top);
			box = cv::Rect(left, top, width, height);
			for (int j = 0; j < keypoint.size(); j++)
			{
				keypoint[j] = ptr[5 + j];
			}
		}
		else if(m_algo_type == YOLO26)
		{
			int left = int(ptr[0]) > 0 ? int(ptr[0]) : 0;
			int top = int(ptr[1]) > 0 ? int(ptr[1]) : 0;
			int width = int(ptr[2] - ptr[0]) > 0 ? int(ptr[2] - ptr[0]) : 0;
			int height = int(ptr[3] - ptr[1])> 0 ? int(ptr[3] - ptr[1]) : 0;
			width = (left + width) < m_image.cols ? width : (m_image.cols - left);
			height = (top + height) < m_image.rows ? height : (m_image.rows - top);
			box = cv::Rect(left, top, width, height);
			for (int j = 0; j < keypoint.size(); j++)
			{
				keypoint[j] = ptr[6 + j];
			}
		}

		boxes.push_back(box);
		scores.push_back(score);
		keypoints.push_back(keypoint);
	}

	scale_boxes(boxes, keypoints, m_image.size());

	if (m_algo_type == YOLOv8 || m_algo_type == YOLOv11 || m_algo_type == YOLOv12)
	{
		std::vector<int> indices;
		nms(boxes, scores, m_score_threshold, m_nms_threshold, indices);
		m_output_pose.clear();
		m_output_pose.resize(indices.size());
		for (int i = 0; i < indices.size(); i++)
		{
			int idx = indices[i];
			OutputPose output;
			output.id = 0;
			output.score = scores[idx];
			output.box = boxes[idx];
			output.keypoint = keypoints[idx];
			m_output_pose[i] = output;
		}
	}
	else if(m_algo_type == YOLO26)
	{
		m_output_pose.clear();
		m_output_pose.resize(boxes.size());
		for (int i = 0; i < boxes.size(); i++)
		{
			OutputPose output;
			output.id = 0;
			output.score = scores[i];
			output.box = boxes[i];
			output.keypoint = keypoints[i];
			m_output_pose[i] = output;
		}
	}

	if(m_draw_result)
		draw_result(m_output_pose);
}
